### PR TITLE
MF-388 Helm Charts for Mainflux

### PR DESCRIPTION
**Pull request by juanmagal**
### What does this do?

It includes the necessary yaml files to deploy Mainflux in a Kubernetes cluster using helm

### Which issue(s) does this PR fix/relate to?
Somehow it lays the foundation stone for issue #388 (Create Helm package for Mainlfux Kubernetes #388)

### List any changes that modify/break current functionality
Some changes were also introduced in the Kubernetes deployment to align with this new functionality. For instance, it includes necessary yaml files to optionally deploy a load balancer (metal-lb) in a bare-metal Kubernetes cluster.

### Have you included tests for your changes?
No, but it was properly tested in a bare-metal Kubernetes cluster. Note Istio ingress was not used, but Nginx.

### Did you document any new/modified functionality?
Yes, including corresponding README files.

### Notes
I am not very used to provide changes using git (git rebase and so on), but using gerrit for that, so there might be things not very well done, my apologies for that.
On the other hand, I am not sure which is the MF Id that I should use.

### CODE REVIEW

File: cmd/users/main.go
Comment (manuio) [inline cmd/users/main.go]: Fix this please

File: docker/nginx.conf
Comment (manuio) [inline docker/nginx.conf]: do we want to change this URL?

**Review by manuio — CHANGES_REQUESTED** [review]

File: cmd/users/main.go
Comment (juanmagal) [reply cmd/users/main.go]: I have fixed it in last push

Comment (manuio) [reply cmd/users/main.go]: Do you mean, in your next commit? Because in this PR there is only one commit.

File: docker/nginx.conf
Comment (juanmagal) [reply docker/nginx.conf]: I need some advice on this change.
So far, in mainflux/k8s/nginx/nginx.conf there is a symbolic link to ../../docker/nginx.conf which means same configuration should be valid for both docker compose and Kubernetes deployments.
In the last commits introduced for Kubernetes  deployments (some weeks ago), specific namespaces were created. For the microservices using grpc it was decided to include them in grpc-services namespace.
Nginx however is belonging to a different namespace (default namespace). Domain names in Kubernetes include the name service together with the namespace. Hence if we have two micro-services belonging to different namespaces, both of them must use the complete domain name to communicate (which is not needed if they belong to the same namespace)
So, I guess there are two options:
1) All microservices are created in the same namespace
2) I remove the symbolic link and from now on there are two different versions of nginx.conf, one for Kubernetes deployment and another one for Docker Compose.

I would suggest the second option, but don't know if you have any other preference

File: cmd/users/main.go
Comment (juanmagal) [reply cmd/users/main.go]: What about now?

Comment (manuio) [reply cmd/users/main.go]: Now it's ok 👍

**Comment by drasko** [comment]
The title of this PR should be `MF-388 Helm Charts for Mainflux`

File: docker/nginx.conf
Comment (manuio) [reply docker/nginx.conf]: @juanmagal  I think that 2 is the good option. The other one will brake the docker composition. To reproduce do `make run` from root.

**Review by manuio — CHANGES_REQUESTED** [review]

File: docker/nginx.conf
Comment (juanmagal) [reply docker/nginx.conf]: Ok, I have removed the link and created two different nginx.conf file: one for docker composition and another one for k8s. Please, check my last changed pushed.

File: docker/nginx.conf
Comment (juanmagal) [reply docker/nginx.conf]: In the last k8s pull request (previous to my pull request) it was decided to have three different namespaces: grpc-services, ingress-nginx and default.
Things, Users and the different adapters are declared under grpc-services and the rest of services (like dashflux) are created under the default namespace. Ingress-nginx was created, but it was not finally used.
So just in case someone decides to use nginx and associate it to ingress-nginx namespace, I included the default namespace for dashflux. That way, if someone creates nginx under ingress-nginx namespace, it will work without having to change nginx.conf; and if he/she decides to create it under default namespace (as it is today), it will also work (even the information is redundant)

Comment (juanmagal) [reply docker/nginx.conf]: Note now that I have created a new nginx.conf file for k8s deployment, "http://dashflux.default:3000/" will be only applicable to k8s and not to docker-compose.
The explanation of why "default" is included is shown in my comment above and it is related to a previous pull request.

**Comment by juanmagal** [comment]
@drasko @manuio just wanted to know if you have any info about this PR

File: k8s/mainflux/ws.yml
Comment (manuio) [inline k8s/mainflux/ws.yml]: @drasko @janko-isidorovic I'm not sure that we want to set min replicas as 1, right?

**Review by manuio — CHANGES_REQUESTED** [review]

File: k8s/mainflux/ws.yml
Comment (juanmagal) [reply k8s/mainflux/ws.yml]: I turned it back to 3 since I used 1 for my tests (I don't have a big system), if you think this number must be changed, I can put whatever you think we should have there.

**Comment by nmarcetic** [comment]
@juanmagal Please update branch, its out of date.

**Comment by exocode** [comment]
Could it be, that this PR **was not merged because of the conflicting `.gitignore` file?** 

This .`gitignore` contains nothing more than the requester's IDE exclusions.

It would be **awesome** if someone of you guys remove or accept this change.
It wont change any behaviour.

Thank you in advance

**Comment by exocode** [comment]
@juanmagal Very well done, awesome! I am trying your PR for a jumpstart of Mainflux for Kubernetes.
I am a Rancher2 user [very nice GUI orchestration Tool for Kubernetes] and try to extend your submission by adding a question/answer form, which is supported by Rancher by simply adding a `questions.yml` you can see it here (https://github.com/exocode/mainflux/blob/mf_helm_1/helm/mainflux/questions.yml).

Independent to this questions.yml I struggle a bit when I do `helm install --namespace <whatever> mainflux/`. Some of the services stay on "Pending"-status. Is this normal behavior? Or did I miss something? (Ingress,...)

Bildschirmfoto 2019-03-10 um 12 59 53

If you can point me in the right direction, we maybe can achieve finishing this PR :-)

Thank you in advance

**Comment by juanmagal** [comment]
Hi,

 There seems to be some problem for the "pending" services. I would need to
see the logs of the different PoDs in order to see which might be the
problem behind. Do you have this info?

Best regards,

Juanma

**Comment by exocode** [comment]
Only the loadbalancers are "pending". For those are no logs available. The pods are all ok: 
I only added an ingress to expose the dashflux ui for port 3000

Maybe this is normal behavior, I am not the k8s geek. I am using bare metal servers, so no GKE or AWS. 
```
> kubectl get service -n mainflux
NAME                       TYPE           CLUSTER-IP      EXTERNAL-IP   PORT(S)                                     AGE
coap-adapter               LoadBalancer   10.43.106.63    <pending>     5683:30801/TCP                              33m
dashflux                   LoadBalancer   10.43.152.159   <pending>     3000:30691/TCP                              33m
http-adapter               LoadBalancer   10.43.118.163   <pending>     8185:32002/TCP                              33m
influxdb-reader            LoadBalancer   10.43.116.85    <pending>     8905:32377/TCP                              33m
influxdb-writer            LoadBalancer   10.43.100.161   <pending>     8900:30869/TCP                              33m
mainflux-nginx             LoadBalancer   10.43.15.83     <pending>     80:31129/TCP,443:30120/TCP,8883:30943/TCP   33m
mainflux-redis             ClusterIP      10.43.33.94     <none>        6379/TCP                                    33m
mainflux-things-postgres   ClusterIP      10.43.166.246   <none>        5432/TCP                                    33m
mainflux-users-postgres    ClusterIP      10.43.138.123   <none>        5432/TCP                                    33m
mqtt-adapter               LoadBalancer   10.43.134.26    <pending>     1883:31497/TCP,8880:32013/TCP               33m
normalizer                 ClusterIP      10.43.133.18    <none>        8184/TCP                                    33m
things                     LoadBalancer   10.43.32.145    <pending>     8182:31347/TCP,8183:30710/TCP               33m
users                      LoadBalancer   10.43.38.7      <pending>     8180:32542/TCP,8181:31963/TCP               33m
ws-adapter                 LoadBalancer   10.43.74.232    <pending>     8186:31456/TCP                              33m
> kubectl describe service -n mainflux influxdb-reader
Name:                     influxdb-reader
Namespace:                mainflux
Labels:                   component=influxdb-reader
                          io.cattle.field/appId=mainflux
Annotations:              <none>
Selector:                 component=influxdb-reader
Type:                     LoadBalancer
IP:                       10.43.116.85
Port:                     <unset>  8905/TCP
TargetPort:               8905/TCP
NodePort:                 <unset>  32377/TCP
Endpoints:                10.42.2.32:8905,10.42.3.13:8905,10.42.4.27:8905
Session Affinity:         None
External Traffic Policy:  Cluster
Events:                   <none>
> kubectl describe service -n mainflux dashflux
Name:                     dashflux
Namespace:                mainflux
Labels:                   component=dashflux
                          io.cattle.field/appId=mainflux
Annotations:              <none>
Selector:                 component=dashflux
Type:                     LoadBalancer
IP:                       10.43.152.159
Port:                     <unset>  3000/TCP
TargetPort:               3000/TCP
NodePort:                 <unset>  30691/TCP
Endpoints:                10.42.1.20:3000,10.42.2.34:3000,10.42.4.24:3000
Session Affinity:         None
External Traffic Policy:  Cluster
Events:                   <none>
```
Here you have an overview how it looks in my Rancher UI:

**Comment by juanmagal** [comment]
Hi,

 The services which are pending are the ones exposed externally with a
routable IP Address. Those are the ones whose type is LoadBalancer.

 When defining a LoadBalancer service there must be a real Load Balacer in
front of these service. This load balancer eill provide the External IP and
will do NATing towards the cluster IPs. When using GKE, Amazon or Azure,
this load balancer is provided natively by the infrastructure, so those
external IPs are automatically assigned.

 However, when it is a custom K8s cluster, you have to deploy a load
balancer in your infrastructure. I deployed  MetalLB to act as load
balancer in my custom cluster. Here there is an example of how to do it:

https://medium.com/@JockDaRock/kubernetes-metal-lb-for-on-prem-baremetal-cluster-in-10-minutes-c2eaeb3fe813

 Best regards,

Juanma

**Comment by drasko** [comment]
@juanmagal NginX in Mainflux system plays the role of a Load Balancer as well (not only reverse proxy). Can this be achieved with existing NginX instead of extrnal LB? Can it be achieved in that way that it is configurable, and that we let people use - either use Mainflux NginX for LB, or external one?

**Comment by juanmagal** [comment]
Hi Drasko,

 Nginx or Istio can be used as an Ingress hiding Things, Users, Adapters
and even Dashflux interfaces behind.

 In the helm charts provided by this PR (and in the existing Kubernetes cli
commands to deploy Mainflux in K8s) Nginx is not configured using Ingress
and Ingress Controller Kubernetes concepts, but it is working as an
Ingress.

 However, as far as I understand, this Ingress still needs an External IP.
When using bare-metal K8s clusters there are different options you can use,
one of them is using a Load Balancer like Metal LB, another one is using a
Node Port (using an ephemeral port mapped to the microservice IP addresd
and port). This is what I found in Nginx documentation and I decided using
Metal LB:

https://kubernetes.github.io/ingress-nginx/deploy/baremetal/

 However, when deploying Nginx in Google Cloud it is also associated to a
Load Balancer service, but Load Balancer is provided by GKE:

https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/provider/cloud-generic.yaml

 For more information see:

https://kubernetes.github.io/ingress-nginx/deploy/#gce-gke


 Best regards,

Juanma

**Comment by exocode** [comment]
I have another question: How can I run `mainflux-cli provision test` to run a initial creation of a user? Which container should I use to `make`the cli? Thank you in advance

**Comment by drasko** [comment]
@exocode we actually build and run CLI on a localhost, so it can interact with the remote Mainflux cloud (from your laptop for example).

If you think that it would be useful to have it in the K8 network, maybe we can consider putting CLI in a container as well...

Is there something  actually preventing you using CLI from your localhost (it uses Mainflux SDK and acts as a Mainflux app, using HTTP REST API)?

**Comment by exocode** [comment]
Ok, clear, but which service I should connect to? mainflux-http?

**Comment by exocode** [comment]
I am still in the setup testing and struggle with the `mainflux-cli provision test` command from my local Mac to the `http-mainflux` workload in the k8s cloud (exposed by a xip.io address via an ingress).
This is the command I used, it's exact that one I use on my local machine too (local I use the docker-machine IP of course):

```
$ mainflux-cli -m https://mainflux-cli.mainflux.116.202.24.192.xip.io provision test
failed to create entity
```

I assigned and tried **ports 8081, 8181, 443, 80**. Are these the right ports?
Am I missing something? Keys, secrets, passwords?

How else would it possible to get my first user if not like this?

**Comment by exocode** [comment]
Ok. What I didn't know was, that I have to assign port 80 (443) the `mainflux-nginx` and not to the `mainflux-http`. Now it works. Regarding the "securing the API": I have to care myself to secure it (it was explained to me in the Gitter chat.

**Comment by drasko** [comment]
@exocode `provision test` has to send requests to Users and Things service, as well as to HTTP adapter service. These are different machines, having different IP address and ports. Because `provision test` command does not know to which addresses and ports live your machines, it can only shoot at port `80` and hope that you have NginX running there which is capable to proxy adequate requests to adequate services behind.

Alternatively - it can work with `docker-compose` because all services map ports to `localhost`, so it actually uses `localhost` as a destination.

However - when in Kubernets, you do not have that and you must run NginX proxy and let CLI use port `80` for this particular command.

**Comment by nicklasfrahm** [comment]
@juanmagal I think you might want to remove the file `helm/mainflux/templates/.normalizer.yaml.swap` as I think that you added it by accident.

Additionally, I have a conceptual question. It seems like we are provisioning multiple services of type LoadBalancer to publish our various protocol connector services. This seems impractical, if we would like to have a single IP for all protocols of our mainflux instance. Even more so, if we want all of our services to be available on the same FQDN.

So my idea was the following: Couldn't we use the nginx for all protocols via the [nginx stream](https://docs.nginx.com/nginx/admin-guide/load-balancer/tcp-udp-load-balancer/) feature. This way we would only need to expose the nginx via one service of type LoadBalancer. Especially when shipping this to cloud environments, such as AWS, Azure or GCP, multiple IPs and load balancers could come at a significant cost.

**Disclaimer:** I do not want to block you in any way and this is also not something that I have tested before.

**Comment by juanmagal** [comment]
There are two swap files that should be removed. I will do it once I have
some time. I guess this is not really a stopper for the testing you are
doing.

Regarding your comment about having multiple services of type of
LoadBalancer, I agree there should not be so many LoadBalancer services. In
fact, when deploying Mainflux in a commercial platform (like Azure or GKE),
I usually set most of the services to "ClusterIP" instead of
"LoadBalancer", since on one hand getting external IP Address is expensive
as you say, and on the other hand it might cause security problems.
However, this is what I found in the k8s configuration yaml files (
https://github.com/mainflux/mainflux/tree/master/k8s) and I don't know if
you guys had special reason for that.

Finally, you guys are also talking about nginx as the ingress for the k8s
deployment, however k8s configuration is also configured to use Istio. I
don't know if you have any preference on that.

**Comment by nicklasfrahm** [comment]
I think we should align with the core maintainers on a strategy here. From the architecture drawing in the documentation I could see that the intention is to use nginx as a load balancer for all given protocols. I have checked with the `docker-compose.yaml` and there it seems like not all protocols are going through the nginx.

If the general idea or intention behind the architecture is outlined and described a bit more in detail, I could submit a PR with an updated or an additional architecture drawing that makes it clear on any infrastructure how to set the application up.

**Comment by drasko** [comment]
It should be noted that with introducing of https://github.com/mainflux/mainflux/pull/676 AuthN will depend on NginX. So we really need this one as an ingress, and proxy/LB for all protocols.

**Comment by nicklasfrahm** [comment]
@drasko If this is the case, we should have a look at the available options for NginX in a Kubernetes environment. To my current knowledge, the options are:
- Creating our own NginX image that manages all incoming traffic.
- Using a namespaced [nginx-ingress](https://github.com/kubernetes/ingress-nginx) and applying some scripts to make `mTLS` work.

**nginx-ingress** might be the better choice as soon as you start deploying your own microservices in the same namespace as **mainflux**. However, I do not know to which extent **nginx-ingress** is extendable, such as the standard NginX.

**Comment by tony-ward** [comment]
Hey guys, my team and I are looking to get a production server running mainflux, probably using kubernetes.  Can somebody give me an update on where this thread led?

**Comment by drasko** [comment]
@tony-ward we are running k8s in production, but not via this Helm chart. It will not be hard to clean this one and send PR, but we are currently under the water - if you have some bandwith please proceed and send PR. @janko-isidorovic took this task, so you can ping him directly.

**Comment by nmarcetic** [comment]
@juanmagal We can't merge this PR, due to inactivity caused by us. It was sent 6 months and now it's impossible to merge. Thank you for your help and contribution, we highly appreciate and you're helping will be logged and mentioned.

**Pull request closed without merging**
